File: python/divide_conquer/sort_list.py
from typing import List, Optional

class ListNode:
    def __init__(self, val: int = 0, next: 'ListNode' = None):
        self.val = val
        self.next = next

    # Helper to represent the list node for easy visualization
    def __repr__(self):
        return f"{self.val} -> {self.next}" if self.next else f"{self.val}"


class Solution:
    def sort_list(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return head

        n = 0

        current = head

        while current:
            n += 1
            current = current.next

        dummy = ListNode(0)
        dummy.next = head
        step = 1

        while step < n:
            prev, curr = dummy, dummy.next

            while curr:
                left = curr
                right = self.split(left, step)

                curr = self.split(right, step)

                prev.next = self.merge(left, right)

                while prev.next:
                    prev = prev.next

            step *= 2

        return dummy.next

    def split(self, head: Optional[ListNode], step: int) -> Optional[ListNode]:
        for _ in range(1, step):
            if head:
                head = head.next
            else:
                return None

        next_segment = head.next

        head.next = None

        return next_segment

    def merge(self,left: Optional[ListNode], right: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(0)
        current = dummy

        while left and right:
            if left.val < right.val:
                current.next = left
                left = left.next
            else:
                current.next  = right
                right = right.next

            current = current.next

        current.next = left if left else right

        return dummy.next

    def print_list(self, head: Optional[ListNode]):
        while head:
            print(head.val, end=" -> " if head.next else "\n")
            head = head.next

    def create_list(self, arr: List[int]) -> Optional[ListNode]:
        dummy = ListNode()

        current = dummy

        for val in arr:
            current.next = ListNode(val)
            current = current.next

        return dummy.next

File: golang/divide_conquer/sort_list.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func sortList(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}

	n := 0

	current := head

	for current != nil {
		n++
		current = current.Next
	}

	dummy := &ListNode{Next: head}

	step := 1

	for step < n {
		prev, curr := dummy, dummy.Next
		for curr != nil {
			left := curr

			right := split(left, step)

			curr = split(right, step)

			prev.Next = merge(left, right)

			for prev.Next != nil {
				prev = prev.Next
			}
		}

		step *= 2
	}

	return dummy.Next
}

func split(head *ListNode, step int) *ListNode {
	for i := 1; head != nil && i < step; i++ {
		head = head.Next
	}

	if head == nil {
		return nil
	}

	next := head.Next

	head.Next = nil

	return next
}

func findMiddle(head *ListNode) *ListNode {
	slow, fast := head, head
	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}
	return slow
}

func merge(left, right *ListNode) *ListNode {
	dummy := &ListNode{}
	curr := dummy

	for left != nil && right != nil {
		if left.Val < right.Val {
			curr.Next = left
			left = left.Next
		} else {
			curr.Next = right
			right = right.Next
		}

		curr = curr.Next
	}

	if left != nil {
		curr.Next = left

	} else {
		curr.Next = right
	}

	return dummy.Next
}

func printList(head *ListNode) {
	for head != nil {
		fmt.Print(head.Val, " ")
		head = head.Next
	}
	fmt.Println()
}

func createList(arr []int) *ListNode {
	dummy := &ListNode{}

	curr := dummy

	for _, val := range arr {
		curr.Next = &ListNode{
			Val: val,
		}

		curr = curr.Next
	}

	return dummy.Next
}
